### DISCUSSION TITLE
WWDC 2023 - Swift Macros and Data

### DISCUSSION THREAD

**Discussion by TC72** [pointfreeco/swift-composable-architecture#2179] [General]
Just wondering how everyone feels about the huge announcements at WWDC?

I laughed a lot to see them use macros which left just environment and state. Obviously these are brand new so nobody's had a chance to do much with them but I think they could be a huge boost to implementing systems like TCA.

Making use of Macros is Swift Data, the promise of simple data storage for Swift UI apps without getting your hands dirty with core data and cloudkit. This is handled by a macro so hopefully if there are any parts of it which aren't compatible with TCA it'll be possible to change the macro to suit.

What does everyone think? I'm just off to sell a kidney ready for the launch of Vision Pro in the UK some time in 2025. 😂

P.S. We'll be getting more info on both of these at sessions today:

![CleanShot 2023-06-06 at 08 10 32](https://github.com/pointfreeco/swift-composable-architecture/assets/402357/5145d662-ac2e-4d17-bfc7-f27c2b5cbee8)

**Comment by mariob** [comment]
Yeah, I'm really excited to try out SwiftData. If you only need a struct to conform to Codable to be able to persist and do queries, that would be awesome.

The whole macro thing seems to open up really great ways to simplify and make things more ergonomic. I was smiling when they had a macro to create a URL from a string without having to put an '!' at the end 😀. I also wonder if macros can make TCA even more ergonomic, for example when it comes to scoping, it would be nice to just have to specify the child feature datatype, but not sure if that's possible to do with a macro.

I really like that they have an 'Observable' annotation now. The view can use any of the properties from such class and the change notification will only come for those properties that the view is actually consuming. I wonder if TCA can use this to get rid of the 'observe'  param you need to specify.

I already sold my kidney yesterday before I say the availabliity date 🤣

**Comment by fabstu** [comment]
How do you distinguish between persistent state and navigation state?

I've got that issue already with having a feed of items (persisted) that I have to scope into using 3 levels of SwitchStore. Would that work better using Data?

**Comment by TC72** [comment]
The Tuesday sessions are available, including what's new in Swift, Swift UI and sessions for macros and swift data:
https://developer.apple.com/wwdc23/sessions/

**Comment by TC72** [comment]
I'm not sure if Swift Data will be compatible with TCA out of the box. After watching Apple's videos it seems like using `@Model` to use Swift Data automatically uses its new observable system. 

`SwiftData supports the all-new observable feature for your modeled properties. SwiftUI will automatically refresh changes on any of the observed properties.`

This automatically does things like publish all the variables and handles changes being propagated to views.

My understanding of TCA isn't very good so I might be getting this wrong. This chart from [their video on observation](https://developer.apple.com/wwdc23/10149) shows how data can be used with observable:

![CleanShot 2023-06-07 at 07 02 03](https://github.com/pointfreeco/swift-composable-architecture/assets/402357/f2a7c1ac-0a08-49af-bb26-ab0fa384074e)

**Reply by kamcma** [reply]
My initial thought is that neither `@Observable` nor `@Model` will be immediately, directly applicable from the perspective of users of TCA. ~~Both require their attached declarations to be classes, not value types.~~ [Edit: apparently not true of `@Observable`, so I retract that.]

In the case of `@Model`, this means those models are not appropriate to use directly in TCA state, which must be value types. Now, if you like the SwiftData API, you could still implement your persistence client with SwiftData, but you would need to copy your data out of `@Model` reference types before returning them to into TCA state, and you'd be knowingly forfeiting much of the change-tracking stuff that comes with these kinds of ORMs. (ORMs are, in general, very effectful frameworks and are an impedance mismatch, or least quite constrained, in unidirectional UI architectures like TCA.)

In the case of `@Observable`, the _library_ itself may be able to use that to improve the implementation of `ViewStore`, which is currently an `ObservableObject`. ~~But again, not something you as a user would mark your state with, because your state must be value types.~~

**Reply by lightandshadow68** [reply]
`@Observable` could alleviate issues with updating unreferenced state with `WithViewStore` in SwiftUI Views. I don't think it would remove the need for the `observe:` parameter, as there are use cases beyond performance. And you can still might want to slice off a bit of state, as a feature could have several collections of row data. 

SwiftData `@Query` , etc. exposes persistence details to the view. Those details are usually implemented in Effects called from a Reducer.  

Unfortunately, SwiftData doesn't support returning query results as structs. And, it's unclear how we could "add" a value type to a context, beyond implementing `Identifiable` and passing in a different value type with the same ID to update. Uniquing, etc. simply wouldn't make sense. Then again, things like uniquing can enable an entire class of bugs. Normally, I'd consider leaking stored models outside an effect an anti-pattern. But if they were effectively identical as a value types after being transformed using `@Model` this might be acceptable with trivial models. 

While SwiftData doesn't provide this, it might make developing a framework that does around it easier than CoreData.

**Comment by arasan01** [comment]
I personally think that a closer look at the Swift Macros announcement could easily create a good way to write Shared State. The current TCA CaseStudies include a Shared State that handwrites getter and setter. This is a straightforward approach but quickly becomes intolerable boilerplate. This is exactly where Macros comes in!

**Reply by TC72** [reply]
Yes, I think TCA could make use of macros to make it even more powerful and easier to work with.

**Reply by mariob** [reply]
Yes, and it would be great if the macro could cache the value because sometimes the getter is called quite often and it could be "expensive" depending how you assemble the child state from the parent state.

**Comment by TC72** [comment]
The latest session on creating an App with Swift Data is available:
https://developer.apple.com/wwdc23/10154

An another looking into schema:
https://developer.apple.com/wwdc23/10195

I am so torn right now. I've been spending a few months learning TCA. I have one App on the app store already built with Swift and then updated for Swift UI. I have ideas for another 4 apps but they will all rely on using iCloud for data storage.
I like TCA as It's just my son and myself working on the apps and the TCA flow along with testing could help us a lot, but Swift Data with @Model and @Query look like they could take care of a lot of complicated data storage for us.
I haven't seen any examples of using TCA with Core Data and CloudKit, I need to work out what's involved.

I think macro libraries will be open to work on so it might be possible to have a TCA implementation of Swift Data that uses the backend parts without the observable parts.

**Comment by mbrandonw** [comment]
Hi everyone, sorry for the delay in response. We were busy this week recording new episodes. 

We’ve spent a little bit of time exploring some of the new tools, but it is far too early to make any concrete statements. There are a lot of exciting things to consider. For one, macros are certainly going to [help out the case path](https://twitter.com/pointfreeco/status/1666870224573075456) situation for us, though we still do hope someday Swift has first class support for the concept.

Macros are also going to be applicable to dependency design and dependency management, parsing, and a whole bunch more.

And then of course we want to be able to support things like `@Observable` and `@Model` in TCA. We don’t have any concrete details, but hopefully `@Observable` alleviates the need for view stores. And at the very least `@Model` is probably immediately usable in dependencies, but using directly in state is going to of course be tricky because it is largely still built on reference types.

So, we hope everyone can be patient as we digest things and explore, and we also hope everyone shares their own explorations to see what is possible. For what it’s worth, this is pretty similar how things went last WWDC (20222) when `NavigationStack` was announced. A discussion was started [the first day](https://github.com/pointfreeco/swift-composable-architecture/discussions/1136) it was announced and it wasn’t until just last week that we released the final version of navigation stacks that works with TCA.

I’m sure there are going to be a lot of explorations, a lot of false starts, and a lot of re-thinking of things. At the end of the day TCA is all about writing applications in a consistent manner with modularity, effects management, testing, and ergonomics in mind. These are the principles we want to always build upon, and let the library evolve however it needs to live up to those ideals and play nicely in Apple’s ecosystem.

**Reply by Alex293** [reply]
Awesome news about CasePaths !!

**Reply by mbrandonw** [reply]
We found some time to explore things yesterday, and as we hoped it does seem like `@Observable` will [largely alleviate](https://twitter.com/pointfreeco/status/1667248728862871612) the need for view stores. And the few places where view stores may be handy we think can further be eliminated with macros. There's a lot still [up in the air](https://forums.swift.org/t/second-review-se-0395-observability/65261/58), but hopefully everything works out! 

We also have some promising experiments with SwiftData. We will share more later.

![ezgif-4-b48c3fa271](https://github.com/pointfreeco/swift-composable-architecture/assets/135203/34b98b30-120b-42f0-93b7-c2bc06ad42f5)

**Reply by acosmicflamingo** [reply]
Can't wait to integrate this in 4 years when my minimum deployment target will be iOS 17.0 LOL

**Reply by mbrandonw** [reply]
We're not sure how much will deploy to iOS <17, but certainly some of it will. :)

**Reply by acosmicflamingo** [reply]
Something is better than nothing :D

**Reply by Camji55** [reply]
Macros themselves should be supported as long as you compile with Xcode 15 and Swift 5.9; observation on the other hand seems to require iOS 17+.

https://developer.apple.com/documentation/Observation?changes=latest_minor

**Reply by acosmicflamingo** [reply]
Observation also seems to be SwiftUI-specific and not for UIKit :( maybe in a couple of years SwiftUI will be production-ready ;)

**Reply by Camji55** [reply]
Check this out! It should still work with UIKit, it's just not as convenient to use.

https://developer.apple.com/documentation/observation/withobservationtracking(_:onchange:)?changes=latest_minor

**Reply by acosmicflamingo** [reply]
@Camji55 fascinating!

**Reply by Camji55** [reply]
@acosmicflamingo here's an example repo with Observability working in UIKit (Still bound by iOS 17+) https://github.com/Camji55/Observation-UIKit/

Syntactically, this isn't ideal but thank god we have macros to abstract some of this away if we wanted 😉
```swift
private func render() {
    withObservationTracking {
        countLabel.text = "\(counter.count)"
    } onChange: { [weak self] in
        Task { @MainActor [weak self] in
            self?.render()
        }
    }
}
```

**Reply by acosmicflamingo** [reply]
@Camji55 WHOA! That's NICE :) Thanks so much for sharing!!!!!! Oh absolutely; so many cool things coming our way!!!

**Reply by fabstu** [reply]
Small question about the SwiftUI example: Does Observable re-render the whole body like ObservableObject, or just the sub-views that used the variable that changed? I'm asking because I thought apart from observing less state, WithViewStore was also used to only re-render that small part of the view instead of the whole view body.

**Reply by mbrandonw** [reply]
@fabstu `@Observable` does cause the whole body to render, but I think in general that is ok. While _technically_ you could hyper localize `WithViewStore` to a small part of the view, I think in practice that didn't happen often and probably didn't make that much of a difference.

**Reply by fabstu** [reply]
Thanks for doing the research! Love ur library. So.. when I have views 5 layers deep, even then it doesn't matter? Just so I understand you well. I get pretty confused with such things. Or I guess there's always the option to factor out any access into their own sub-views.

**Reply by mbrandonw** [reply]
`WithViewStore` matters in that it helps the view from rendering more than it needs to, just like `@Observable`. But, `WithViewStore` also allows you to hyper focus where the observation happens, like this:

```swift
var body: some View {
  NavigationStack {
    VStack {
      WithViewStore(…) { viewStore in 
        // Something that uses viewStore
      }
    }
  }
}
```

The fact that `ViewStore` was brought deeper into the view isn't really saving that much work. Sure you don't recreate `NavigationStack` and `VStack` again, but that's probably doesn't really matter.

That hyper focus is the feature of `WithViewStore` that `@Observable` can't replicate, but we don't think that is a big deal.

**Reply by fabstu** [reply]
So if below the view its 10-20 layers deep or when I have a sole TabView below that takes long to re-render, then probably all top-layer access should go into its own view instead 🤔. Thanks for the clarification!
I feel like I use TCA a bit wrong now. Everything in my hobby project is scoped like that, hmm.

**Reply by mbrandonw** [reply]
I'm not sure I understand your comment enough to say that is correct or not. 

The short of it is that pretty much the best you can do in TCA right now is to define a `ViewState` struct for each view that only holds the state your view cares about, and then install `WithViewStore` at the root. That is if you want to be really diligent.

With `@Observable` you will be able to forget the `ViewState` and the `WithViewStore` and it will function the same as the diligent style of today.

**Reply by fabstu** [reply]
Except that it re-renders all views embedded in the view body as well. If that's not an issue, sure. But if it would make the UI flicker, then its just a new paradigm to move all access into their own views to avoid it. I guess I'm too pessimistic. I loved everything u guys do so far, so I'm not worried. :)

**Reply by mbrandonw** [reply]
I'm sorry I'm just not following. There is no flickering problem, and I'm not sure what moving  things to another view accomplishes.

Let's just leave it with the fact that TCA will behave more like vanilla SwiftUI, and that `WithViewStore` wasn't doing any magic to prevent needless re-renders. It simply did `removeDuplicates` on state changes to decide when to trigger a `body` recompilation.

**Reply by SouzaRodrigo61** [reply]
I'm studying and implementing an application using composable architecture and SwiftData, and I've created an initial sample repository on TCA + SwiftData. If anyone wants to help or improve this approach, I appreciate it.

https://github.com/SouzaRodrigo61/SwiftDataTCA

**Reply by ghost** [reply]
Imho, TCA and SwiftData would work fine together, if the changes submitted by `@Query` would find their way to a feature's state, so that the reducer can react on it. 

Would it be possible to use the `.bind` property (in a variant) for achieving this? Or are there any other ideas?

**Reply by DandyLyons** [reply]
I'm using the same approach that @SouzaRodrigo61 started in their SwiftDataTCA repo. Not using `@Query` at all. 

We create a swift data dependency and a dependency for each model and that's all of the communication between TCA and SwiftData. No need for a `.bind`. 

I'm liking it so far.

**Reply by mbrandonw** [reply]
SwiftData should work in TCA in the sense that changes to a SwiftData model will cause the view to re-render even though it's a reference type, which prior to Observation was not the case. But, testing SwiftData models in state is going to be very tricky. You can't perform exhaustive testing since the models are reference types. At best you can only assert on individual fields of the model after an action is sent.

**Reply by SouzaRodrigo61** [reply]
@DandyLyons I'm building an application following the `SwiftDataTCA` concept that I developed, and it's been working well so far. Without the need for `@Query`, all queries are performed using the SwiftData concept for UIKit.

https://github.com/SouzaRodrigo61/cashew

In a part of the code, I have data retrieval done like this:

```swift
return .run { send in
    let data = try loadData()
    
    await withTaskGroup(of: Void.self) { group in
        group.addTask {
            await send(.taskResponse(.success(data)), animation: .default)
        }
        
        group.addTask(priority: .background) {
            await self.startTimer(send: send)
        }
    }
} catch: { error, send in
    await send(.taskResponse(.failure(error)), animation: .default)
}
```

For the fetch part, I do it like this. In my case, I don't have much need to add various validations, and I leave all control flow to TCA.

```swift
fetch: {
    do {
        @Dependency(\.modelContext.context) var context
        let movieContext = try context()
        
        let descriptor = FetchDescriptor<Task.Model>()
        return try movieContext.fetch(descriptor)
    } catch {
        return []
    }
},
```

**Reply by GierlJn** [reply]
Are there any new ideas on how we can get the changes submitted by the `@Query` macro to TCA State?

**Reply by stephencelis** [reply]
@GierlJn `@Query` only works in views, so you could add an `onChange(of: self.query)` to feed updates into TCA.
